# Section 3: Series  Methods and Handling
This section will cover the following topics:


*   series structure, item count, NaNs (aka NA's or nulls, which are common features of real-world data)
*   descriptive statistics
*   series arithmetics
*   sorting series
*   filtering and custom transformations





In [1]:
import pandas as pd
import numpy as np

## Reading in Data with `read_csv()`

Pandas has a powerful method called `read_csv()` that can be used to read csv files into dataframes.
*   `read_csv()` has over a dozen different possible parameters, but most of the time you won't need to specify most of them
*   Documentation: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html



Start by reading in the file **drinks.csv** using `read_csv()`. If the file is local, you would pass the file path to the `read_csv()` method. However, if the file is hosted remotely, you can simply pass the URL to the file.

In [2]:
drinks_df = pd.read_csv('https://gist.githubusercontent.com/sh7ata/e075ff35b51ebb0d2d577fbe1d19ebc9/raw/b966d02c7c26bcca60703acb1390e938a65a35cb/drinks.csv')
drinks_df

country  beer_servings  ...  wine_servings  total_litres_of_pure_alcohol
0    Afghanistan            NaN  ...            NaN                           NaN
1        Albania           89.0  ...           54.0                           4.9
2        Algeria           25.0  ...           14.0                           0.7
3        Andorra          245.0  ...          312.0                          12.4
4         Angola          217.0  ...           45.0                           5.9
..           ...            ...  ...            ...                           ...
188    Venezuela          333.0  ...            3.0                           7.7
189      Vietnam          111.0  ...            1.0                           2.0
190        Yemen            6.0  ...            NaN                           0.1
191       Zambia           32.0  ...            4.0                           2.5
192     Zimbabwe           64.0  ...            4.0                           4.7

[193 rows x 5 columns]

This object that we've created is NOT a series - rather, it is a dataframe. We will cover dataframes extensively in the next section.
For now, let's pare this down and select just the **country** and **wine_servings** columns. We can specify the columns using the `usecols` parameter

In [3]:
wine = pd.read_csv('https://gist.githubusercontent.com/sh7ata/e075ff35b51ebb0d2d577fbe1d19ebc9/raw/b966d02c7c26bcca60703acb1390e938a65a35cb/drinks.csv', usecols=['country','wine_servings'])
wine

country  wine_servings
0    Afghanistan            NaN
1        Albania           54.0
2        Algeria           14.0
3        Andorra          312.0
4         Angola           45.0
..           ...            ...
188    Venezuela            3.0
189      Vietnam            1.0
190        Yemen            NaN
191       Zambia            4.0
192     Zimbabwe            4.0

[193 rows x 2 columns]

Next, let's set the country column as the index. This can be done using the `index_col` parameter in the `read_csv()` call. This will replace the integer indices with the country names!
Alternatively, we could have used the `.set_index()` method which allows you to choose a column of an existing dataframe to change set as the index. More info here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_index.html?highlight=set_index#pandas.DataFrame.set_index
The code to do this would have been:
`wine.set_index('country')`

In [4]:
wine = pd.read_csv('https://gist.githubusercontent.com/sh7ata/e075ff35b51ebb0d2d577fbe1d19ebc9/raw/b966d02c7c26bcca60703acb1390e938a65a35cb/drinks.csv', usecols=['country','wine_servings'], index_col = 'country')
wine

wine_servings
country                   
Afghanistan            NaN
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0
...                    ...
Venezuela              3.0
Vietnam                1.0
Yemen                  NaN
Zambia                 4.0
Zimbabwe               4.0

[193 rows x 1 columns]

In [5]:
wine.head()

wine_servings
country                   
Afghanistan            NaN
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0

Note that even though we only have one column of values and its associated labels (country), this is technically still a dataframe and NOT a series. This is because `read_csv()` always returns a dataframe from a CSV. You'll notice that it looks visually different from a series as well, with the column labels positioned differently.


*   However, you can use the `squeeze` parameter in `read_csv` in order to return a dataframe if the parsed data contains only one column. `squeeze` is set to `False` by default




In [6]:
type(wine)

pandas.core.frame.DataFrame

In [7]:
wine = pd.read_csv('https://gist.githubusercontent.com/sh7ata/e075ff35b51ebb0d2d577fbe1d19ebc9/raw/b966d02c7c26bcca60703acb1390e938a65a35cb/drinks.csv', usecols=['country','wine_servings'], index_col = 'country', squeeze = True)
wine

country
Afghanistan      NaN
Albania         54.0
Algeria         14.0
Andorra        312.0
Angola          45.0
               ...  
Venezuela        3.0
Vietnam          1.0
Yemen            NaN
Zambia           4.0
Zimbabwe         4.0
Name: wine_servings, Length: 193, dtype: float64

`wine` is now a Series instead of a dataframe

In [8]:
type(wine)

pandas.core.series.Series

## Series Sizing with `.size`, `.shape`, and `len()`

Recall that **attributes** give you properties of the series, whereas **methods** perform actions. In this lecture we'll explore series attriburtes

The **size** attribute gives you the length of the series. In the example below, we see that the `wine` series has 193 observiations, for a length of 193.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.size.html

In [9]:
wine.size

193

We can also check the sizes of the series values and index components separately. Using the `.values` attribute, we get an array containing the values for the series.

In [10]:
wine.values

array([ nan,  54.,  14., 312.,  45.,  45., 221.,  11., 212., 191.,   5.,
        51.,   7.,  nan,  36.,  42., 212.,   8.,  13.,  nan,   8.,   8.,
        35.,  16.,   1.,  94.,   7.,  nan,   7.,  16.,   1.,   4.,   1.,
         1.,   1., 172.,   8.,   3.,   1.,   9.,  74.,  11., 254.,   5.,
       113., 134.,  nan,   1., 278.,   3.,  26.,   9.,   3.,   1.,   2.,
       233.,  nan,  59.,  nan,   1.,  97.,  37.,  59.,   1., 149., 175.,
         1., 218.,  28.,   2.,   2.,  21.,   1.,   1.,   2., 185.,  78.,
        nan,  nan,  nan,  nan, 165.,   9., 237.,   9.,  16.,   1.,  12.,
         2.,   1.,  nan,   6., 123.,  62.,  31.,  nan,   2.,  nan,  56.,
       271.,   4.,   1.,  nan,  nan,   1.,  12.,  nan,  nan,  18.,   5.,
        18.,  nan,   8., 128.,   1.,   5.,  nan,   1.,   8.,  nan,  19.,
       175.,   1.,   1.,   2.,   7., 129.,   1.,  nan,  23.,  18.,   1.,
        74.,  21.,   1.,  56., 339.,   7.,   9.,  18., 167.,  73.,  nan,
        32.,  71.,  11.,  24.,  nan,  14.,  nan,   

The values array actually has its own `.size` attribute.

In [11]:
wine.values.size

193

Similarly, we use the `.index` attribute to get an array of the indices. It too has its own `.size` attribute

In [12]:
print(wine.index)
print('The size of wine.index is {}'.format(wine.index.size))

Index(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua & Barbuda', 'Argentina', 'Armenia', 'Australia', 'Austria',
       ...
       'Tanzania', 'USA', 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Venezuela',
       'Vietnam', 'Yemen', 'Zambia', 'Zimbabwe'],
      dtype='object', name='country', length=193)
The size of wine.index is 193


The `.shape` attribute returns a *tuple* that provides the dimensions of the series or dataframe in (rows, columns). Since series are one-dimensional, the shape will be a tuple with just one element.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.shape.html

In [13]:
print(wine.shape)

(193,)


Finally, the built-in `len()` function from Python will also give you the size of the series
* https://docs.python.org/3/library/functions.html#len

In [14]:
print(len(wine))

193


## Unique Values and Series Monotonicity

The `is_unique` attribute is an intreresting attribute that checks whether a series contains a sequence of unique values. That is, it checks whether every value in the series is different.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.is_unique.html

For our wine series, the series in its entirety is not unique because two or more of the values are identical. However, if we take just the first five values, those are unique.

In [15]:
print(wine.is_unique)
print(wine.head(5).is_unique)

False
True


The `nunique()` method returns the number of unique values in the series.
* Note that `nunique()` disregards 'NA' values by default. You can set the `dropna` parameter to True in order to include it. 
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.nunique.html

In [16]:
print(wine.nunique())
print(wine.nunique(dropna = False))

71
72


The `is_monotonic` attribute returns a boolean that determines if the values in the object are monotonic_increasing. 
* **Monotonicity** refers to whether the series values are always increasing (or staying the same) or always decreasing (or staying the same)
* Keep in mind that the `is_monotonic` attribute is actually an alias for the `monotonic_increasing` attribute.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.is_monotonic.html



In [17]:
print(pd.Series([1,2,3]).is_monotonic)
print(pd.Series([1,2,2,2,3,10,100]).is_monotonic)
print(pd.Series([1,2,1,1,3]).is_monotonic)
print(pd.Series([1,2,1,1,3, -10]).is_monotonic)


True
True
False
False


How do we use `is_monotonic` to test for decreasing monotonicity? To do this, we can wrap `is_monotonic` in a `reversed()` method.
* https://docs.python.org/3/library/functions.html#reversed

In [18]:
print(pd.Series([10,9,8,7,6,5,4,3,2,1]).is_monotonic)
print(pd.Series(reversed([10,9,8,7,6,5,4,3,2,1])).is_monotonic)

False
True


## The `count()` Method

The `count()` method is a convenient method that allows you to count the values in your series. What makes it different from `.size` is that `.count()` returns the **total number of non-NA/non-null values**, whereas `.size` includes those values.
* NA or null values are like gaps in the data, indicating absence of information
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.count.html

In [19]:
print(wine.size)
print(wine.count())

193
162


We can make this more formal by using the `.hasnans` attribute to determine whether a series has any `nans`

In [20]:
print(wine.hasnans)

True


## Accessing and Counting NAs/Nulls
* Nulls and NAs indicate the absence of data

The `isnull()` method is used on a Series to return a same-sized object indicating if the corresponding values are NA values
* NA valeus include None, NaN, NA, and Null
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.isnull.html#pandas.Series.isnull

In [21]:
wine.isnull()

country
Afghanistan     True
Albania        False
Algeria        False
Andorra        False
Angola         False
               ...  
Venezuela      False
Vietnam        False
Yemen           True
Zambia         False
Zimbabwe       False
Name: wine_servings, Length: 193, dtype: bool

Great, we have a Series of booleans indicating whether that value is Null. But that's not terribly helpful. How do we get the actual countries instead of a Series of booleans?
We can use the output of `isnull()` as a **boolean mask** to get the actual countries!


In [22]:
wine[wine.isnull()]

country
Afghanistan        NaN
Bangladesh         NaN
Bhutan             NaN
Burundi            NaN
North Korea        NaN
Eritrea            NaN
Ethiopia           NaN
India              NaN
Indonesia          NaN
Iran               NaN
Iraq               NaN
Kuwait             NaN
Lesotho            NaN
Libya              NaN
Malaysia           NaN
Maldives           NaN
Marshall Islands   NaN
Mauritania         NaN
Monaco             NaN
Myanmar            NaN
Nepal              NaN
Pakistan           NaN
Rwanda             NaN
San Marino         NaN
Saudi Arabia       NaN
Somalia            NaN
Sri Lanka          NaN
Sudan              NaN
Tajikistan         NaN
Uganda             NaN
Yemen              NaN
Name: wine_servings, dtype: float64

What we got back from the above call was a new series in which only the countries without wine serving data are selected. However, for that reason, the Series values of NaN are not very useful. So if we want only the country names, we can use an index call.

In [23]:
wine[wine.isnull()].index

Index(['Afghanistan', 'Bangladesh', 'Bhutan', 'Burundi', 'North Korea',
       'Eritrea', 'Ethiopia', 'India', 'Indonesia', 'Iran', 'Iraq', 'Kuwait',
       'Lesotho', 'Libya', 'Malaysia', 'Maldives', 'Marshall Islands',
       'Mauritania', 'Monaco', 'Myanmar', 'Nepal', 'Pakistan', 'Rwanda',
       'San Marino', 'Saudi Arabia', 'Somalia', 'Sri Lanka', 'Sudan',
       'Tajikistan', 'Uganda', 'Yemen'],
      dtype='object', name='country')

Going one step further, we can transform this into a regular Python list by using the `list()` constructor. This list can then be used however you'd like

In [24]:
list(wine[wine.isnull()].index)

['Afghanistan',
 'Bangladesh',
 'Bhutan',
 'Burundi',
 'North Korea',
 'Eritrea',
 'Ethiopia',
 'India',
 'Indonesia',
 'Iran',
 'Iraq',
 'Kuwait',
 'Lesotho',
 'Libya',
 'Malaysia',
 'Maldives',
 'Marshall Islands',
 'Mauritania',
 'Monaco',
 'Myanmar',
 'Nepal',
 'Pakistan',
 'Rwanda',
 'San Marino',
 'Saudi Arabia',
 'Somalia',
 'Sri Lanka',
 'Sudan',
 'Tajikistan',
 'Uganda',
 'Yemen']

How do we count null values in a series? One way is to just find the length of the list that we just generated. This works, but it's unwieldy because we first had to isolate the nulls in a series by obtaining a boolean mask, then apply that mask to the series, and then find the length of the series isolated by the boolean mask.

In [25]:
len(list(wine[wine.isnull()].index))

31

A more *pandorabe* (adorable and involving pandas) way to do this is to chain the `isnull()` method with the `sum()` method. 
* Normally, the `sum()` method will sum all of the values in a series. However, when operating on a boolean series, it will treat *True* as 1 and *False* as 0, which is true for Python in general. Therefore, the sum will simply be the sum of the number of `True` values, which in this case corresponds to the values that are Null.

In [26]:
print(pd.Series([True, False, True]).sum())

print(wine.isnull().sum())

2
31


Finally, let's ensure the count of values by type is fully accounted for. That is, the null and non-null values should add up to the total size of the series. Let's test this!

In [27]:
all = wine.size

In [28]:
nonnulls = wine.count()

In [29]:
nulls = wine.isnull().sum()

In [30]:
all == nonnulls + nulls

True

A final note is that `isnull()` and `isna()` are aliases of each other. You can use either.

## Another Approach for Isolating Nulls
Might look more elegant. It uses a Numpy universal function (`ufunc`), which allows us to perform vectorized operations.
* Vectorized operations work on an entire collection at once, as opposed to applying the operation to one element of a collection at a time
* It is not recommended to use `ufunc` directly because they can be very picky and finicky
* https://numpy.org/doc/stable/reference/ufuncs.html#available-ufuncs
* https://numpy.org/doc/stable/reference/generated/numpy.isnan.html#numpy.isnan

Let's create a simple series of booleans, a null value, and an integer.



In [31]:
ser = pd.Series(data = [True, False, None, 2], dtype = float)
print(ser)

0    1.0
1    0.0
2    NaN
3    2.0
dtype: float64


No, use the `np.isnan()` function on this series. What returns is a numpy array of booleans that describes whether each value in the series is or is not a null (NaN) value
* Booleans are NOT Null values, nor is the integer `2`

In [32]:
np.isnan(ser)

0    False
1    False
2     True
3    False
dtype: bool

Instead of passing a Pandas series to the `np.isnan()` function, we could also take a reference of a numpy ufunc and pass it directly to the series.
* This is indexing by a callable function! It will essentially test every value in the `wine` series using the `np.isnan()` function, and will only select the values that return as **True** from the function.

In [33]:
print(wine[np.isnan])
print(wine[np.isnan].size)

country
Afghanistan        NaN
Bangladesh         NaN
Bhutan             NaN
Burundi            NaN
North Korea        NaN
Eritrea            NaN
Ethiopia           NaN
India              NaN
Indonesia          NaN
Iran               NaN
Iraq               NaN
Kuwait             NaN
Lesotho            NaN
Libya              NaN
Malaysia           NaN
Maldives           NaN
Marshall Islands   NaN
Mauritania         NaN
Monaco             NaN
Myanmar            NaN
Nepal              NaN
Pakistan           NaN
Rwanda             NaN
San Marino         NaN
Saudi Arabia       NaN
Somalia            NaN
Sri Lanka          NaN
Sudan              NaN
Tajikistan         NaN
Uganda             NaN
Yemen              NaN
Name: wine_servings, dtype: float64
31


## The `notnull()` and `notna() Methods

The `notnull()` method returns a series of booleans indicating whether each record is NOT Null
* A Null value will give False
* A non-Null value will give True
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.notnull.html

In [34]:
wine.notnull()

country
Afghanistan    False
Albania         True
Algeria         True
Andorra         True
Angola          True
               ...  
Venezuela       True
Vietnam         True
Yemen          False
Zambia          True
Zimbabwe        True
Name: wine_servings, Length: 193, dtype: bool

This can be used as a boolean mask to index the non-nulls only

In [35]:
wine.loc[wine.notnull()]

country
Albania               54.0
Algeria               14.0
Andorra              312.0
Angola                45.0
Antigua & Barbuda     45.0
                     ...  
Vanuatu               11.0
Venezuela              3.0
Vietnam                1.0
Zambia                 4.0
Zimbabwe               4.0
Name: wine_servings, Length: 162, dtype: float64

And we can use `notnull()` combined with the `.sum()` method to add up the number of True's (basically add up the count of values that are not null)

In [36]:
wine.notnull().sum()

162

To prove that `notnull()` and `isnull()` methods are complementary, let's add their results together to ensure they add up to the total length of the series (spoiler: they do)

In [37]:
wine.notnull().sum() + wine.isnull().sum() == wine.size

True

The `notnull()` method has an aliased method called `notna()`

*   https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.notna.html



## Booleans are Literally Numbers in Python!
* `True` is simply a keyword for the integer `1`
* `False` is simply a keyword for the integer `0`


In [38]:
True + 19

20

In [39]:
True + True - False + (True * 3)

5

Just like in regular division, you cannot divide by zero

In [40]:
5 / False

ZeroDivisionError: ignored

Another way to say this is that the `bool` types *inherit* from the `int` class. Inheritance is a facet of object-oriented programming. If we examine the method resolution order (MRO), we will see that it looks for methods first within the `bool` object type, next within the `int` object type, and finally within the `object` object type

In [ ]:
print(type(True))
print(type(False))
bool.__mro__

This explains why we are able to call the `sum()` method, which operates on integers, on a list of boolean values and it will work!

## Dropping and Filling NAs

The `dropna()` method simply returns a **new** series with the NA values dropped
* Important: it does NOT alter the original series UNLESS the `inplace` argument is set to True, or if the variable name for the original series is reassigned to the new series
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dropna.html

In [ ]:
wine.dropna()

Another approach to dealing with NAs is to replace them with something that is more meaningful or computational useful. This is accomplished with the `fillna()` method. This method requires passing in a value that you want to replace the NA values with.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.fillna.html

In [ ]:
print(wine)
print(wine.fillna(100, inplace=False))

## Descriptive Statistics Using Series
* Pandas features several methods that allow you to calculate useful descriptive statistics for data Series

To find the **average** value of your series, you can calculate it manually by using the `sum()` method and then dividing the result by  the `count()`

In [ ]:
average_manual = wine.sum() / wine.count()
average_manual

A better and faster way to do it by using the `mean()` method.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.mean.html

In [ ]:
wine.mean()

You can also calculate the median value (the value at the very middle of a list of numbers, or the 50th percentile), using the `median()` method
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.median.html

In [ ]:
wine.median()

Note that in the above example, the median is much smaller than the mean. This signifies that most of the values are below the mean, and there are just a handful of very large values that skew the mean upward.

Another way to say this is that the wine servings are right-skewed, or positively skewed. 
* Right-skewed (or positively skewed) means that the distribution of the data has a drawn out tail on the right, and most of the data is on the left
* Left-skewed (or negatively skewed) means that the distribution of the data has a drawn out tail on the left, and most of the data is on the right
* https://en.wikipedia.org/wiki/Skewness


Let's visualize the data to see what it actually looks like? We'll do more with data viz later, but for now we'll just use the Pandas built-in `hist()` function to construct a histogram of the series. As we can see , most countries have between 0 and 50 wine servings, and the data is heavily right-skewed

In [ ]:
wine.hist()

We can also calculate quantiles using the `quantile()` method!
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.quantile.html

In [ ]:
print(wine.quantile(0.5))
print(wine.quantile(0.25))
print(wine.quantile(0.75))

We can calculate the **interquartile range** (or IQR), which is the difference between the 75th and 25th percentiles

In [ ]:
iqr = wine.quantile(0.75) - wine.quantile(0.25)
iqr

Pandas has built-in `min()` and `max()` functions which, as you imagine, return the smallest and largest values of the series, repsectively
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.min.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.max.html

In [ ]:
print(wine.min())
print(wine.max())

The built-in functions `var()` and `std()` allow you to calculate the variance and standard deviation of the data, respectively. Both provide a measurement of how widely the data are distributed around the mean.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.std.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.var.html

In [ ]:
print(wine.std())
print(wine.var())
print(wine.std() ** 2 == wine.var())

## The `describe()` Method

The `describe()` method generates a set of descriptive statistics, including the count, mean, standard deviation, min, max, first, and third quartiles
* It is recommended that this be one of the first things you do with a set of data
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.describe.html

In [ ]:
wine.describe()

The method is customizable using the arguments. For example, we can specify the percentiles that we want to calculate


In [ ]:
wine.describe(percentiles = [0.79, 0.19])

The `include` and `exclude` parameters allow you to choose which types of data that will be included into or excluded from the `describe` calculations

In [ ]:
wine.describe(include = float, exclude = object)

## `mode()` and `value_counts()`

The **mode** is a descriptive statistic that defines the most commonly-occurring value(s) within your data distribution. In Pandas, the `mode()` method can be used to obtain the mode value(s)
* Datasets can have multiple modes (multimodal)
* Drops NA values by default, but this can be modified with the `dropna` parameter
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.mode.html

In [ ]:
wine.mode()

We can calculate how frequently the mode (or any other value for that matter) appears in the dataset. One way to do this is with a boolean mask, selecting only for countries whose value is 1.

In [ ]:
print(wine == 1)
print('Count of Countries with 1 Wine Serving')
print(wine[wine == 1].count())

A faster way of accomplishing this is to use the `valuecounts()` method
* This method returns a series that describes the number of occurrences for each unique value
* By default, the returned series is sorted by descending order, NAs are dropped, and the series is not normalized
* It includes a normalization parameter
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html

In [ ]:
wine.value_counts()

To get the mode, you can simply select the first item in the `value_counts()` series

In [ ]:
wine.value_counts().iloc[0]

When the **normalize** paramter is used, the returned values are ratios of the total count (relative frequencies), as opposed to absolute counts 

In [ ]:
wine.value_counts(normalize=True)

Let's prove that this is the case with a simple calculation using `value_counts()` and `count()`:

In [ ]:
wine.value_counts().iloc[0] / wine.count()

## `idxmax()` and `idxmin()`

The `idxmax()` method returns the *row label* that corresponds to the maximum value in the series. 
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.idxmax.html

In [ ]:
wine.idxmax()

An alternative way of doing this is to use a logical operator with square bracket indexing in combination with the `max()` function. However, this will return a series instead of a row label.

In [ ]:
print(wine[wine == wine.max()])

In order to access the actual country, we can append the `.index` attribute to return the index object, and then use index notation to access the first item in that object.

In [ ]:
print(wine[wine == wine.max()].index)
print(wine[wine == wine.max()].index[0])

The inverse method for finding the row label corresponding to the minimum value is `indxmin()`
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.idxmin.html

In [ ]:
wine.idxmin()

One must take care with these methods because there can be multiple row labels that correspond to the minimum and maximum values. For example, recall from earlier that 28 different countries have a wine serving number of 1 (the minimum value).

More specifically, the methods will return the *alphabetically first* row label corresponding to the query.

In order to get ALL row labels corresponding to the minimum value, we can use a logical test with bracket notation:

In [ ]:
wine[wine == wine.min()]

As a brief side note, if we use the `idxmax()` and `idxmin()` methods with bracket indexing, it will return the actual value instead of the row label. In that sense, it functions just like `min()` or `max()`

Why does this occur? recall that the `idxmax()` and `idxmin()` methods return the *row label* of the value that corresponds to the desired value. When passing a row label into bracket indexing OR into `loc`, that returns the *value* of the row corresponding to that label.

In [ ]:
wine[wine.idxmax()]

## Sorting with `sort_values()`

Generally speaking, Pandas series can be sorted by **values** or by **index**. 
To sort by values, we use the `sort_values()` method
* By default, the series will be sorted in ascending order but can be changed with the `ascending` parameter
* It places the NaNs at the end of the sorted series by default, but they can be placed at the beginning with the `na_position` parameter
* We can also specify the sorting algorithm, where we can use `quicksort`, `mergesort`, or `heapsort`
 * These algorithms have different performance implications depending on the nature of the data, its structure, and how sorted it is to begin with
 * The author does not recommend changing this parameter in most cases, as `quicksort` is usually appropriate
* Similar to other methods, by default `sort_values()` returns a copy of the series as sorted. The original series is not affected. This can be changed using the `inplace` parameter.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.sort_values.html

In [ ]:
print(wine.sort_values())
print(wine.sort_values(ascending = False))
print(wine.sort_values(na_position='first'))

## `nlargest()` and `nsmallest()`


What if we wanted to know the `n` largest or smallest values in a series?

One way to do this would be to sort the values in either ascending or descending order, and then using the `head()` or `tail()` method or bracket indexing as appropriate.


In [ ]:
print(wine.sort_values(ascending=False).head(10))
print(wine.sort_values(ascending=False)[:11])

An easier way to do this is the use the `nlargest()` or `nsmallest()` methods, which accomplish the same task with the same
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.nlargest.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.nsmallest.html

In [ ]:
print(wine.nlargest(10))
print(wine.nsmallest(35))

## Sorting with `sort_index()`


Similarly to sorting by values, we can also sort by the index labels using the `sort_index()` method.
* By default, the series is sorted by index alphanumerically in ascending order, placing NA values at the end. Here we are of course referring to NAs in the index, not in the values.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.sort_index.html

In [ ]:
wine.head()

In [ ]:
print(wine.sort_index())

In [ ]:
print(wine.sort_index(ascending=False))

## Series Arithmetic and `fill_value()`

Basic mathematics can be conducted on Pandas series. For example, simple addition can be used to add a number to all non-NaN values in the series:

In [ ]:
print(wine)
print(wine + 2)

Another trivial and meaningless example:

In [ ]:
(wine - 10) * 2

Where Pandas really shines is in **juggling and combining multiple series**
In Pandas, indices align automatically.

Let's start be creating a sorted series of our wine servings data

In [ ]:
series1 = wine.sort_index()
series1

We now create a complementary series to use alongside this. 

In [ ]:
more_drinks = pd.Series({'Albania' : 6})
more_drinks

What we will try to do now is to combine the two Series so that 6 drinks are added to Albania in `series1`

In [ ]:
series1 + more_drinks

What happened here? 6 drinks *were* added to **Albania**, but all of the other values went to to NaN. This is because **Albania** was the only key that is in both `series1` and `more_drinks`. That's just how the addition of series works.

To drive this point home, let's add drinks to another country as well.

In [ ]:
more_drinks = pd.Series({'Albania' : 6, 'Algeria': 10})
more_drinks

In [ ]:
series1 + more_drinks

This is clearly not convenient, as we lose the drinks info from all of the countries where no drink addition was made. What we want to do is to add the drinks from both countries together and keep the info from both series. 

Instead of using of using the addition operator, we instead use the series `add()` method, which adds two series together element by element.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.add.html

In [ ]:
series1.add(more_drinks)

But wait, this doesn't look any different from when we used the addition operator! Remember that **series addition goes element-by-element**. In this case, more_drinks is missing most of the elements that `series1` has, and thus when they are added, the result for the missing elements is NaN. 

The key here is to use the `fill_value` parameter. What this essential does is fill in all missing values in `more_drinks` with the value that we specify by the `fill_value` parameter, thereby **guaranteeing that the two series have elements that align by index label**. It usually makes sense to use `0` as the fill_value.
* Another way to think of this is that `fill_value` plugs the gaps and makes way for perfect label alignment. 

In [ ]:
series1.add(more_drinks, fill_value=0)

Notice that some elements are `NaN` in both `series1` and in `more_drinks`. In those cases, the addition will still resolve to NaN. 
* If data in both corresponding Series locations is missing the result of filling (at that location) will be missing.

The `subtract()` method offers the same functionality, except with subtraction instead of addition.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.subtract.html

In [ ]:
series1.subtract(more_drinks, fill_value=0)

The `multiply()` and `divide()` methods work in the same manner.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.divide.html
 * For `divide()`, you probably don't want to use 0 as the `fill_value` in order to avoid division by 0 errors
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.multiply.html

In [ ]:
wine.divide(more_drinks, fill_value = 1)

In [ ]:
wine.multiply(more_drinks, fill_value = 1)

Important note: the *order* of the records in each respective series is not important. Once combined within a Pandas method, Pandas will connect them by label prior to applying the applicable operator in the values.

This is one of the key features of working with data in Pandas and it makes life much easier, since you do not need to worry about aligning the data manually between the two series.

## Calculating Variance and Standard Deviation

We've already seen how to calculate the variance and standard deviation using their respective built-in methods.

**Variance** is the average of squared differences from the mean. The further the values are from the mean, the greater their squares and the higher the variance.
The **sample standard deviation** is simply the square root of the variance.
* https://www.learntocalculate.com/how-to-calculate-sample-variance/

In [ ]:
wine.var()

Instead of using the `var()` method, let's try to replicate the variance calculation based on the formula for variance and the arithmetic methods we just learned.

First subtract the mean from all values in `wine` using the `subtract()` method.

In [ ]:
subtract_mean = wine.subtract(wine.mean())
subtract_mean

Then square each of those differences using the `multiply()` method.

In [ ]:
distance_squared = subtract_mean.multiply(subtract_mean)
distance_squared

Finally, calcute the variance by summing all values in `distance_squared` and dividing by the number of (non-NaN) values minus 1.

In [ ]:
variance = distance_squared.sum() / (distance_squared.count() - 1)
variance

We can also combine all of this into a single line as follows:

In [ ]:
variance = wine.subtract(wine.mean()).multiply(wine.subtract(wine.mean())).sum() / (wine.count() - 1)
variance

Another option is to use the power operator `**` instead of `multiply()`

In [ ]:
variance = (wine.subtract(wine.mean())**2).sum() / (wine.count() - 1)
variance

## Cumulative Operations

**Cumulative operations** are those that return cumulative values along a series or data frame.

The `cumsum()` method returns a series in which each subsequent value is the cumulative sum of all values up to and including that point.
* By extension, the final entry will be the total that is returned by `sum()`.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cumsum.html

In [ ]:
print('Total sum: {}'.format(wine.sum()))
print(wine.cumsum())

By default, NaN values are skipped. If the `skipna` parameter is set to `False`, then any sum that includes an NaN value will resolve to NaN.


In [ ]:
pd.NA + 1000

The `prod()` method, gives the total product of all values in the series multiplied together
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.prod.html


In [ ]:
wine.prod()

By contrast, the `cumprod()` method returns a series that reports the cumulate product.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cumprod.html

In [ ]:
wine.cumprod()

The `cummin()` and `cummax()` methods work similarly, which trace the minimum and maximum values (respectively) up until that point in the series or dataframe
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cummin.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cummax.html

In [ ]:
wine.cummin()

In [ ]:
wine.cummax()

## Calculating Pairwise Differences with `diff()`

`diff()` is an extremely useful method that is used to calculate the discrete difference between subsequent pairs of elements in a series. That is, for any given entry in the series, the method returns the difference of that value minus the value of an entry before it.
* By default, the **period** (the number of values that you "go back" to when calculating the difference is equal to 1. However, these can be adjusted using the `periods` parameter.
* Where it truly shines is in data science with *time series*, where you are interested in the mangnitude of differences between subsequent entries in the time series
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.diff.html

In [ ]:
wine.head(5)

In [ ]:
wine.diff().head()

## Series Iteration

Occasionally, we'll find ourselves wanted to iterate over each element in a series, similar to running a *for loop*. 

In fact, we can use a for loop to iterate over the *values* in a series

In [ ]:
mini_wine = wine[:10]
mini_wine

In [ ]:
for i in mini_wine:
  print(i)

What if we want to iterative over the indices? Simple enough, just use the `index` attribute on the series. 

In [ ]:
for i in mini_wine.index:
  print(i)

What if we wanted to access BOTH the indices AND the values at the same time?

One way to do this in pure Python is to iterate over each index, and then use that index label and square bracketing (or `loc`) to also access that value

In [ ]:
for i in mini_wine.index:
  print(i, mini_wine[i])

A more Pandorable way of doing this is to use the `items()` method to iterate over each "item" in the series.
* Each "item" consists if an (index, value) tuple. note that this is different from the string output from the Pythonic methods
* It is a lazy iteration, meaning that it is only evaluated when accessed
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.items.html

In [ ]:
for i in mini_wine.items():
  print(i)

An alias for the `items()` method is `iteritems()`. The two methods work exactly the same way.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.iteritems.html

## Filtering Series: `filter()`, `where()`, and `mask()`

The `filter()` method, as its name implies, allows you to filter you Series based on specified index labels.
* One useful feature of this method is the `regex` paramter, which allows you to use regular expressions (http://www.rexegg.com/regex-quickstart.html) to filter for particular labels based on regex patterns.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.filter.html

In this simple example, the `filter()` method is used to filter the Series for any countries that start with "V"

In [ ]:
wine.filter(regex = "^V")

Another paramter in the `filter()` method is the `like` parameter. It functions similarly to the "like" operator in SQL. It filters for labels that contain the specified string.

Keep in mind that when using the `filter()` method, we always filter along the index labels, NOT the values.
* Recall that you can filter by values using several means, including square bracketing, `loc`, or using callables.

In [ ]:
wine.filter(like='stan')

The `where()` method is not specifically designed for filtering. Instead, it allows us to replace values in a series where a given condition is False
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.where.html

In this example, any value that causes the conditional x > 200 to resolve to "False" (that is, any value 200 or smaller) will be replaced by the string "too small" using the `other` parameter
* If `other` is not specified, it defaults to `NaN`

In [ ]:
wine.where(lambda x: x > 200, other='too small')

If we were to keep the `NaN` default, we could apply the `dropna()` method to get rid of the values that resolved to False in the above `where()` statement

In [ ]:
wine.where(lambda x: x > 200).dropna()

Note that this is equivalent to filtering by square bracketing

In [ ]:
wine[wine > 200]

In all of these examples, we are have filtered by values that satisfy the condition. How would we filter AGAINST values that satisfy the condition?

One way to do this is to invert the criteria. Borrowing from our example above, if we take the inverse of x > 200, namely x <= 200, then all values ABOVE 200 will be replaced with `NaN`. These can then be dropped, giving us all values 200 or below.

In [ ]:
wine.where(lambda x: x <= 200).dropna()

Sometimes, however, logic cannot be inverted so easily. For such cases, when you  are interested in filtering values for conditions that are FALSE, we use the `mask()` method. Similar to `where()`, `mask()` is a replacement method. However. It operates on the exact opposite principle: whereas `where()` replaces values where the condition is False, `mask()` replaces values where the condition is True.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.mask.html

In this example, any values that are above 200 will be replaced by `NaN` (by default) and will then be dropped from the series.

In [ ]:
wine.mask(lambda x: x > 200).dropna()

## Transforming Series with `update()`, `apply()`, and `map()`
Transformations in Pandas fit into two broad catagories:
* **Spot transformations** target specific records directly, for example if you want to switch everything of one value to a different value
* **Global transformations** have a much broader scope, and generally apply to the entire Series

We could always rely on direct assignment to do a transformation:

In [42]:
wine.loc['Algeria'] = 19
wine.head(10)

country
Afghanistan            NaN
Albania               54.0
Algeria               19.0
Andorra              312.0
Angola                45.0
Antigua & Barbuda     45.0
Argentina            221.0
Armenia               11.0
Australia            212.0
Austria              191.0
Name: wine_servings, dtype: float64

However, this gets extremely repetitive if you want to change multiple values. It does not scale well at all. So what do we do?

One option is to use the `update()` method, which modifies a Series *in place* by using values from a different series that is passed in. 
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.update.html

In [45]:
wine.update(pd.Series(data=[200, 20], index=['Albania', 'Algeria']))
wine

country
Afghanistan      NaN
Albania        200.0
Algeria         20.0
Andorra        312.0
Angola          45.0
               ...  
Venezuela        3.0
Vietnam          1.0
Yemen            NaN
Zambia           4.0
Zimbabwe         4.0
Name: wine_servings, Length: 193, dtype: float64

What if we intend to apply a common transformation to *all* elements in the series?

There are several options, but one of the most flexible is the `apply()` method. This method applies a transformation to every single element in the series.
* Essentially you define a function and pass that function into the `apply()` method. The function will then be applied to all elements.
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.apply.html

In this example, each value in the Series is squared (raised to the second power)

In [48]:
wine.apply(lambda x: x**2)

country
Afghanistan        NaN
Albania        40000.0
Algeria          400.0
Andorra        97344.0
Angola          2025.0
                ...   
Venezuela          9.0
Vietnam            1.0
Yemen              NaN
Zambia            16.0
Zimbabwe          16.0
Name: wine_servings, Length: 193, dtype: float64

Another way to do this would be to use a numpy universal function. One such example if the `np.square` function

In [51]:
wine.apply(np.square)

country
Afghanistan        NaN
Albania        40000.0
Algeria          400.0
Andorra        97344.0
Angola          2025.0
                ...   
Venezuela          9.0
Vietnam            1.0
Yemen              NaN
Zambia            16.0
Zimbabwe          16.0
Name: wine_servings, Length: 193, dtype: float64

Yet another option is to define a function externally, then pass it into the `apply()` method.

In [52]:
def square(x):
  return x ** 2

wine.apply(square)

country
Afghanistan        NaN
Albania        40000.0
Algeria          400.0
Andorra        97344.0
Angola          2025.0
                ...   
Venezuela          9.0
Vietnam            1.0
Yemen              NaN
Zambia            16.0
Zimbabwe          16.0
Name: wine_servings, Length: 193, dtype: float64

`apply()` also allows you to pass in arguments to the function in addition to the Series value. This allows you to modify the behavior of the function based on those additional arguments. 

One way to do this is to pass in *positional arguments* (args) with a tuple. The position of the argument in the tuple corresponds to the position in the function definition, after the Series value.

In this example, we'll control what the function does depending on the value that is passed in.

In [58]:
def square_if_less_than_min(x, min_servings):
  if x < min_servings:
    return x ** 2
  return x
print(wine)
print('')
print(wine.apply(square_if_less_than_min, args=(200, )))

country
Afghanistan      NaN
Albania        200.0
Algeria         20.0
Andorra        312.0
Angola          45.0
               ...  
Venezuela        3.0
Vietnam          1.0
Yemen            NaN
Zambia           4.0
Zimbabwe         4.0
Name: wine_servings, Length: 193, dtype: float64

country
Afghanistan       NaN
Albania         200.0
Algeria         400.0
Andorra         312.0
Angola         2025.0
                ...  
Venezuela         9.0
Vietnam           1.0
Yemen             NaN
Zambia           16.0
Zimbabwe         16.0
Name: wine_servings, Length: 193, dtype: float64


Another approach to this is to use keyword arguments (kwarg) that are defined directly within the `apply()` call. The kwargs must match exactly the argument names in the function definition. This has the added advantage of not having to worry about positions. As long as you know what the arguments are called in the function definition, you can use kwargs.


In [59]:
wine.apply(square_if_less_than_min, min_servings = 200)

country
Afghanistan       NaN
Albania         200.0
Algeria         400.0
Andorra         312.0
Angola         2025.0
                ...  
Venezuela         9.0
Vietnam           1.0
Yemen             NaN
Zambia           16.0
Zimbabwe         16.0
Name: wine_servings, Length: 193, dtype: float64

Another way to apply global transformations is with the `map()` method. This is technically a substitution method, not a transformation method. it substitutes values in a Series based on input correspondence. Each value in the Series is substituted with a new value, which may be derived from a function, a dictionary, or another Series
* If the provided argument is a dictionary, an values in the Series that are not in the dictionary are converted into `NaN`
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html

In many cases, `map()` works exactly the same as `apply()`

In [61]:
wine.map(lambda x: x**2)

country
Afghanistan        NaN
Albania        40000.0
Algeria          400.0
Andorra        97344.0
Angola          2025.0
                ...   
Venezuela          9.0
Vietnam            1.0
Yemen              NaN
Zambia            16.0
Zimbabwe          16.0
Name: wine_servings, Length: 193, dtype: float64

However, when we get into more complex transformations (such as functions that accept multiple arguments), `map()` can no longer be substituted for `apply()`. Generally, `apply()` is the more flexible method.

In [65]:
wine.map({20.0:20.4})

country
Afghanistan     NaN
Albania         NaN
Algeria        20.4
Andorra         NaN
Angola          NaN
               ... 
Venezuela       NaN
Vietnam         NaN
Yemen           NaN
Zambia          NaN
Zimbabwe        NaN
Name: wine_servings, Length: 193, dtype: float64